### PR TITLE
adding PrivateBin to Janitor

### PR DESCRIPTION
**Pull request by elrido**
Hello @jankeromnes 

I finally found time to put together a Janitor image for PrivateBin, as outlined in the [Janitor community](https://discourse.janitor.technology/t/how-to-add-a-project/165).

PrivateBin is a PHP & JS web application. While it is simple to set up, we always had complaints (even among the core developers) that the unit testing setup was to complex and therefore many errors only get caught in our pull request triggered travis CI runs and subsequently have to be fixed by other developers, since the requester can't run the test locally.

I really like the idea of Janitor, as a way to provide less savvy contributors an easy way to pop up a PrivateBin development environment, do their changes, run the tests and then create their pull requests.

The container image adds a nginx + php-fpm setup exposing the Git repository, so one can instantly review the result of any changes in the browser. I am not sure if all the GUI stuff is needed for Janitor or not. If you like I can work on a slimmer container that doesn't include VNC etc. For PrivateBin the WebIDEs (which include a terminal to run tests) and the nginx server should be sufficient.

I did build and test the container locally. My application for a Janitor account sent back in March was not yet approved. I'll be glad to test the container in Janitor once I get my account.

For the [SVG logo](https://github.com/JanitorTechnology/janitor/tree/master/static/img) I would propose our [icon.svg from our assets](https://github.com/PrivateBin/assets/blob/master/images/svgs/icon.svg). Should I do a separate pull request for that?

Kind Regards
El RIDO

### CODE REVIEW

File: privatebin/janitor.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    "8090": {
+      "label": "Theia",
+      "proxy": "https"
+    }
+  },
+  "scripts": {
+    "Run all (phpunit & mocha) tests": "unit-test",
+    "Run frontend (mocha) tests": "cd /home/user/privatebin/js && mocha",
+    "Run backend (phpunit) tests": "cd /home/user/privatebin/tst && phpunit",
```
Comment (jankeromnes) [inline privatebin/janitor.json]: Same nit as above about the `&&`.

File: privatebin/privatebin.dockerfile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+# Download deps
+RUN sudo apt-get update && \
+    sudo apt-get -y install --no-install-recommends nginx php-fpm php-gd php-sqlite3 phpunit && \
+    npm install -g mocha jsverify jsdom@9 jsdom-global@2 mime-types && \
+
+# Configure php-fpm
+    sudo mkdir -p /run/php && \
+    sudo sed -ri "s/^;daemonize = .*$/daemonize = no/" /etc/php/7.0/fpm/php-fpm.conf && \
+    sudo sed -ri "s/^(user|group) = .*$/\1 = user/" /etc/php/7.0/fpm/pool.d/www.conf && \
```
Comment (jankeromnes) [inline privatebin/privatebin.dockerfile]: Nit: Oh, are you sure that the multi-line `RUN` trick with `&& \` works when you insert empty lines and comments below? We should check if the CI Docker build complained about this.

File: privatebin/privatebin.dockerfile [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+# Configure supervisord
+    sudo sh -c "echo '[program:php-fpm]\n\
+user = user\n\
+command = sudo /usr/sbin/php-fpm7.0\n\
+autorestart = true\n\
+[program:nginx]\n\
+user = user\n\
+command = sudo /usr/sbin/nginx -g \"daemon off;\"\n\
+autorestart = true' >> /etc/supervisord.conf" && \
```
Comment (jankeromnes) [inline privatebin/privatebin.dockerfile]: Nit: Please feel free to move this extended Supervisor configuration into a separate `supervisord-append.conf` file, like [here](https://github.com/JanitorTechnology/dockerfiles/blob/master/discourse/supervisord-append.conf).

A less hacky way would be to import the file directly into `/etc/supervisor/conf.d/` (e.g. as `/etc/supervisor/conf.d/privatebin.conf`). We might have to add `[include]\nfiles = /etc/supervisor/conf.d/*` to our global [supervisord.conf](https://github.com/JanitorTechnology/dockerfiles/blob/master/ubuntu-dev/supervisord.conf) file for this to work.

File: privatebin/privatebin.dockerfile [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+# Prepare Git repository & clean up
+    git clone https://github.com/PrivateBin/PrivateBin /home/user/privatebin && \
```
Comment (jankeromnes) [inline privatebin/privatebin.dockerfile]: Nit: I feel like this `git clone` could be run in a separate `RUN` command (you don't have to do *all* the setup in a single `RUN` command). This would make this file a bit more readable.

In general it's ok to use as many `RUN` instructions as you want (except for deleting files, this needs to be in the same `RUN` where they're created, otherwise they'll be captured in a Docker image layer).

File: privatebin/privatebin.dockerfile [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+WORKDIR /home/user/privatebin
+
+# Configure the IDEs to use PrivateBin's source directory as workspace.
+ENV WORKSPACE /home/user/privatebin/
+
+# Nginx port
+EXPOSE 80
+
+# Configure Janitor for PrivateBin
+COPY janitor.json /home/user/
```
Comment (jankeromnes) [inline privatebin/privatebin.dockerfile]: Nit: You probably want to add `--chown=user:user` just after the `COPY` keyword, otherwise this file will belong to `root` (that's Docker's default behavior for `COPY` and `ADD`, whatever `USER` is set to).

File: privatebin/privatebin.dockerfile (hunk 2 above)
Comment (elrido) [reply privatebin/privatebin.dockerfile]: I have only tested this on Docker 1.13.1 (the docker.io package from the Ubuntu 16.04 repos). If I switch to the method outlined by you below it will certainly be more robust.

**Comment by elrido** [comment]
I think your suggestions are all very sensible and clear and I'll update the pull request accordingly (and will create a separate one for the SVG).

Regarding putting the Janitor configuration into our main repository: If this would be the preferred, future-proof way, I can of course do that instead. I was trying to move non-standard uses and tools out of the main repository into dedicated ones (i.e. assets, docker for hosting the project, docker for unit testing). Would it be a viable alternative if we create a separate repo just for the Janitor image, too?

**Comment by jankeromnes** [comment]
> I think your suggestions are all very sensible and clear and I'll update the pull request accordingly (and will create a separate one for the SVG).

Perfect, thanks!

> Regarding putting the Janitor configuration into our main repository: If this would be the preferred, future-proof way, I can of course do that instead. I was trying to move non-standard uses and tools out of the main repository into dedicated ones

So, it all depends how frequently you want the development image to be updated.

If you want it to update every time there is a new push to https://github.com/PrivateBin/PrivateBin, the only options I know require some files to be added to that same repository:
- Add a development Dockerfile somewhere in the PrivateBin repository, then log into Docker Hub, create a new Automated Build for the PrivateBin repository (call it e.g. `privatebin/privatebin-dev`), and specify the path to your Dockerfile inside the PrivateBin repository
- Or, add a more flexible Circle CI configuration, that knows where to find your Dockerfile and then push the resulting image to Docker Hub (note: only the CircleCI config needs to be in the PrivateBin repo; the Dockerfile itself can be pulled from somewhere else)

Setting up your Dockerfile in this repository means that your resulting image won't be built on every push to PrivateBin/PrivateBin, but on every push to JanitorTechnology/dockerfiles (sometimes there are weeks without commits).

> Would it be a viable alternative if we create a separate repo just for the Janitor image, too?

So, this would mean that your development image updates everytime there is a new push to that separate repository. I believe that once it's set up, you'll very rarely push to it, so your development image risks getting very much outdated.

TL;DR (sorry, I like to be very chatty sometimes), you have two options:

- Merge this Dockerfile into https://github.com/JanitorTechnology/dockerfiles/ (we're happy and excited to accept it!), and we'll set up your development image as `janitortechnology/privatebin`, and update it as frequently as we can (but sometimes it can get outdated, and you'll need to ping us)
- Merge this Dockerfile into https://github.com/PrivateBin/PrivateBin instead (e.g. in a `.janitor` directory), then set up the development image on Docker Hub yourself (e.g. `privatebin/privatebin-dev` or `privatebin/privatebin-janitor`), and it'll automatically stay up-to-date forever

There might be other options that I don't know of (e.g. using Travis CI if it knows how to build Docker images, or some Docker Hub web hooks).

**Comment by elrido** [comment]
Thank you for your detailed reply.

Given that this is a script project, our development or production environment requirements change rarely. New libraries are added directly to the source code repository and don't need to be deployed to the environment via deb packages. The only downside I see with an outdated image is that a dev always has to run the "Update source code" action in cloud9 before starting to work.

I think for now I would prefer to add the image here and will of course keep an eye on this repository to look out for changes that may be necessary like for ubuntu 18.04 (I think you have already branch for that and I could start to take a look into that).

File: privatebin/janitor.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "scripts": {
+    "Run all tests": "unit-test",
+    "Run frontend tests": {
+      "cmd":  "mocha",
+      "cwd": "/js"
+    },
+    "Run backend tests": {
+      "cmd":  "phpunit",
+      "cwd": "/tst"
```
Comment (jankeromnes) [inline privatebin/janitor.json]: Nit: Shouldn't this be `/test` instead of `/tst`?

**Comment by elrido** [comment]
Thank you!

**Comment by jankeromnes** [comment]
@elrido You're very welcome!

Also, the PrivateBin image build succeeded, and I deployed it to Janitor (Palmaria server): https://janitor.technology/projects/

Could you please try it, and make sure that its containers are usable to hack & test PrivateBin, and to make valuable contributions upstream?

Additionally, you might want to add a `janitor.json` script that starts the PrivateBin server / web app, [like so](https://github.com/JanitorTechnology/janitor/blob/a39c1901a253b6d55d75b17898e2d76d38cd34ed/.janitor/janitor.json#L32).

**Comment by elrido** [comment]
Ok, so here is my feedback (thank you for the invite to the alpha BTW):

Both starting the container via the Janitor website or locally (using `docker run -d --rm -p 127.0.0.1:8080:80 -p 127.0.0.1:8089:8089 -p 127.0.0.1:8090:8090 janitortechnology/privatebin /usr/bin/supervisord`) works well.

When running locally as described above, I can access and use the cloud9 IDE on localhost:8089, Theia IDE on localhost:8090 and the live privatebin preview at localhost:8080.

When running it via the Janitor website, I can launch cloud9 IDE and running git update, the combined or individual unit tests via the menu Run, Run configurations works well (and quite fast, even compared to my local environment). Starting the VNC viewer works and gives you an xterm already in the project folder. I forgot to add any browser, so there is no way to access the live privatebin website (that I have found).

I think that even without the ability to browse the live privatebin website preview, the container in Janitor makes it very easy to change code and run the test suite against it. Would it (in a later version) be possible to configure Janitor images to expose other webservices in the "My Container" view or should I install chromium or firefox in the container image so one can access the preview via the VNC viewer?
